sycl/include/sycl/ext/oneapi/experimental/user_defined_reductions.hpp

```diff
@@ -0,0 +1,119 @@
+namespace ext::oneapi::experimental {
+
+// ---- reduce_over_group
+template <typename GroupHelper, typename T, typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<GroupHelper>), T>
+reduce_over_group(GroupHelper group_helper, T x, BinaryOperation binary_op) {
+  if constexpr (sycl::detail::is_native_op<T, BinaryOperation>::value) {
+    return sycl::reduce_over_group(group_helper.get_group(), x, binary_op);
+  }
+#ifdef __SYCL_DEVICE_ONLY__
+  T *Memory = reinterpret_cast<T *>(group_helper.get_memory().data());
+  auto g = group_helper.get_group();
+  Memory[g.get_local_linear_id()] = x;
+  group_barrier(g);
+  T result = Memory[0];
+  if (g.leader()) {
+    for (int i = 1; i < g.get_local_linear_range(); i++) {
+      result = binary_op(result, Memory[i]);
+    }
+  }
+  group_barrier(g);
+  return group_broadcast(g, result);
+#else
+  std::ignore = group_helper;
+  throw runtime_error("Group algorithms are not supported on host.",
+                      PI_ERROR_INVALID_DEVICE);
+#endif
+}
+
+template <typename GroupHelper, typename V, typename T,
+          typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<GroupHelper>), T>
+reduce_over_group(GroupHelper group_helper, V x, T init,
+                  BinaryOperation binary_op) {
+  if constexpr (sycl::detail::is_native_op<V, BinaryOperation>::value &&
+                sycl::detail::is_native_op<T, BinaryOperation>::value) {
+    return sycl::reduce_over_group(group_helper.get_group(), x, init,
+                                   binary_op);
+  }
+#ifdef __SYCL_DEVICE_ONLY__
+  return binary_op(init, reduce_over_group(group_helper, x, binary_op));
+#else
+  std::ignore = group_helper;
+  throw runtime_error("Group algorithms are not supported on host.",
+                      PI_ERROR_INVALID_DEVICE);
+#endif
+}
+
+// ---- joint_reduce
+template <typename GroupHelper, typename Ptr, typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<GroupHelper> &&
+                           sycl::detail::is_pointer<Ptr>::value),
+                          typename sycl::detail::remove_pointer<Ptr>::type>
+joint_reduce(GroupHelper group_helper, Ptr first, Ptr last,
+             BinaryOperation binary_op) {
+  if constexpr (sycl::detail::is_native_op<
+                    typename sycl::detail::remove_pointer<Ptr>::type,
+                    BinaryOperation>::value) {
+    return sycl::joint_reduce(group_helper.get_group(), first, last, binary_op);
+  }
+#ifdef __SYCL_DEVICE_ONLY__
+  // TODO: the complexity is linear and not logarithmic. Something like
+  // https://github.com/intel/llvm/blob/8ebd912679f27943d8ef6c33a9775347dce6b80d/sycl/include/sycl/reduction.hpp#L1810-L1818
+  // might be applicable here
+  using T = typename std::remove_pointer<Ptr>::type;
+  auto g = group_helper.get_group();
+  T partial = *(first + g.get_local_linear_id());
+  Ptr second = first + g.get_local_linear_range();
+  sycl::detail::for_each(g, second, last,
```

Comment:
Would it work with non-pointer iterators?

Reply:
Probably not, SYCL2020 `sycl::joint_reduce` also uses this approach. User passes sycl::accessor::get_pointer() to this func.

Reply:
At the moment this won't work with non-pointer iterators simply because we have `enable_if` at the beginning of the function, which requires `Ptr` to be a pointer type.

I would suggest that we go with the current approach as a first iteration of the extension: we will anyway only claim it as experimental for now.

@dm-vodopyanov, we should raise this question with spec writers, because both extension and core spec do not say whether `Ptr` can be a generic iterator or is it a mere pointer.